### ISSUE TITLE
Remounting app layout on build of pages

### ISSUE THREAD

**Issue by stramel** [zeit/next.js#4653]
# Bug report

## Describe the bug

When running a custom `_app.js` and using an app-layout method. Navigating to pages that have not been built, cause the server to build. Upon completion of the page build, the app-layout is triggering a re-mount.

In the HMR output, I can see `_app.js` and `layout.js` saying it "changed". Also, through some debugging I found that the chunkHashes are changing on each build for all pages with the exception of `_document.js` and `_error.js`.

I have been working to reproduce the issue but haven't been able to yet.

My `_app.js#render` looks like:

```js
render() {
    const { Component, pageProps, router, store } = this.props
    const page = isFunction(Component.renderLayout) ? (
      Component.renderLayout({
        children: <Component {...pageProps} router={router} />,
        ...pageProps,
        router,
      })
    ) : (
      <Component {...pageProps} router={router} />
    )

    return (
      <Container>
        <Provider store={store}>
          <>
            {/* Apply dynamic tags */}
            <Head />
            {/* Render Persistent Layout */}
            {page}
          </>
        </Provider>
      </Container>
    )
  }
}
```

## To Reproduce (Attempting)

Hopefully I will get it to reproduce. Still unable to reproduce with https://github.com/stramel/next-issue-4653

## Expected behavior
I expect the building of a new page to not cause a re-mount of the app-layout just as it would function in production as a SPA.

## System information
 - OS: MacOS, Windows 10
 - Browser Chrome, Firefox
 - Version of Next.js: 6.0.4-canary.8

**Comment by stramel** [comment]
This is really strange... The auth pages, login/forgot-password/register, all keep the same chunkHash when one of them is built and the `_app` hash doesn't change. However, once I go into an authenticated route, the every hash except `_document` and `_error` change.

```sh
{ chunkHashes:
   Map {
     _c:
      Map {
        'bundles/pages/_app.js' => 'c1b299592a894dcbe0fe3b8e44cbe83c',
        'bundles/pages/_document.js' => 'b6610cffa69fb7970598da62aab9f826',
        'bundles/pages/_error.js' => '3cd548cef2407b419af1d07e39805911',
        'bundles/pages/login.js' => 'f8052e5fe7f53695f8a5e08bfd01b476' } },
  prevChunkHashes:
   Map {
     _c:
      Map {
        'bundles/pages/_app.js' => 'bf16a3e8f154e0807a956dce0b04e3db',
        'bundles/pages/_document.js' => 'b6610cffa69fb7970598da62aab9f826',
        'bundles/pages/_error.js' => '3cd548cef2407b419af1d07e39805911',
        'bundles/pages/login.js' => 'f8052e5fe7f53695f8a5e08bfd01b476' } } }
> Building page: /forgotpassword
 DONE  Compiled successfully in 1061ms
{ chunkHashes:
   Map {
     _c:
      Map {
        'bundles/pages/_app.js' => 'c1b299592a894dcbe0fe3b8e44cbe83c',
        'bundles/pages/_document.js' => 'b6610cffa69fb7970598da62aab9f826',
        'bundles/pages/_error.js' => '3cd548cef2407b419af1d07e39805911',
        'bundles/pages/login.js' => 'bb574c73e5f81e89997fadbcee2aef3e',
        'bundles/pages/forgotpassword.js' => '8afb3c2cb22a027bd076839b15ceb29d' } },
  prevChunkHashes:
   Map {
     _c:
      Map {
        'bundles/pages/_app.js' => 'c1b299592a894dcbe0fe3b8e44cbe83c',
        'bundles/pages/_document.js' => 'b6610cffa69fb7970598da62aab9f826',
        'bundles/pages/_error.js' => '3cd548cef2407b419af1d07e39805911',
        'bundles/pages/login.js' => 'f8052e5fe7f53695f8a5e08bfd01b476' } } }
> Building page: /register
 DONE  Compiled successfully in 637ms
{ chunkHashes:
   Map {
     _c:
      Map {
        'bundles/pages/_app.js' => 'c1b299592a894dcbe0fe3b8e44cbe83c',
        'bundles/pages/_document.js' => 'b6610cffa69fb7970598da62aab9f826',
        'bundles/pages/_error.js' => '3cd548cef2407b419af1d07e39805911',
        'bundles/pages/login.js' => 'bb574c73e5f81e89997fadbcee2aef3e',
        'bundles/pages/forgotpassword.js' => '8afb3c2cb22a027bd076839b15ceb29d',
        'bundles/pages/register.js' => 'a0b13058e387304885a303816457cf42' } },
  prevChunkHashes:
   Map {
     _c:
      Map {
        'bundles/pages/_app.js' => 'c1b299592a894dcbe0fe3b8e44cbe83c',
        'bundles/pages/_document.js' => 'b6610cffa69fb7970598da62aab9f826',
        'bundles/pages/_error.js' => '3cd548cef2407b419af1d07e39805911',
        'bundles/pages/login.js' => 'bb574c73e5f81e89997fadbcee2aef3e',
        'bundles/pages/forgotpassword.js' => '8afb3c2cb22a027bd076839b15ceb29d' } } }
> Building page: /networks/detail
 DONE  Compiled successfully in 4430ms                                                                                                                                                                                                                                                             09:29:16

{ chunkHashes:
   Map {
     _c:
      Map {
        'bundles/pages/_app.js' => '27d1a4a8f3eccc527e9dc3babf034632',
        'bundles/pages/_document.js' => 'b6610cffa69fb7970598da62aab9f826',
        'bundles/pages/_error.js' => '3cd548cef2407b419af1d07e39805911',
        'bundles/pages/login.js' => 'c99b7f006a2ed86eca3e1880f70978f3',
        'bundles/pages/forgotpassword.js' => '65a00390ab7f941b5c4a260bfc0d8572',
        'bundles/pages/register.js' => 'd7bd4ccfd238a9e88a77fb78e730c7df',
        'bundles/pages/networks/detail.js' => '278917fd57eb5b6e51b917b6defbd2e3' } },
  prevChunkHashes:
   Map {
     _c:
      Map {
        'bundles/pages/_app.js' => 'c1b299592a894dcbe0fe3b8e44cbe83c',
        'bundles/pages/_document.js' => 'b6610cffa69fb7970598da62aab9f826',
        'bundles/pages/_error.js' => '3cd548cef2407b419af1d07e39805911',
        'bundles/pages/login.js' => 'bb574c73e5f81e89997fadbcee2aef3e',
        'bundles/pages/forgotpassword.js' => '8afb3c2cb22a027bd076839b15ceb29d',
        'bundles/pages/register.js' => 'a0b13058e387304885a303816457cf42' } } }
```

In this last chunkHash output, I would expect `login, forgot-password, register` page hashes not to have changed. Ideally, `_app` would also not change as well.

**Comment by stramel** [comment]
Alright, getting this narrowed down slowly...

Given 3 pages, `login`, `register`, and `forgot-password`

Starting at `login` navigating to `register`, causes re-mount of `_app` and layout components. Also marks a bunch of files as "changed" which haven't including the login page.

Starting at `login` navigating to `forgot-password` does NOT cause remount of `_app` or layout components. However, it does mark a bunch of files as "changed" which haven't including the login page.
If navigating to `forgot-password` after having visited `register`, it says no files have "changed".

This is where it gets weird IMO.

Starting at `register` and navigating to `login` and then `forgot-password` does NOT cause `_app` or layout components to re-mount. Also, navigating to those pages, shows no files have "changed" in the HMR.

Trying to get this in a minimal repo

**Comment by sourcesoft** [comment]
@stramel Did you have any chance to troubleshoot this? I have the exact same problem. Not just the _app component remounts but also redux store calls @@INIT again.

**Comment by sourcesoft** [comment]
I just found out the issue, not sure if it's supposed to work like this by design.

If a route I navigate to has sub-route like `/page/hello` served by `pages/page.js` it's going to be remounted, however navigating to `/page` or `/about` works just fine.

I had to change `/page/hello` to `/page?q=hello` for now till I found a fix for it. I tried `shallow` routing and `asPath` but none of the works.
